cmd/gardenlet/app/bootstrap.go

```diff
@@ -21,121 +21,65 @@ import (
 	"github.com/gardener/gardener/pkg/client/kubernetes"
 	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
 	"github.com/gardener/gardener/pkg/gardenlet/bootstrap"
-	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	bootstraputil "github.com/gardener/gardener/pkg/gardenlet/bootstrap/util"
 
 	"github.com/sirupsen/logrus"
-	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	certificatesv1beta1client "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
-	"k8s.io/client-go/tools/clientcmd"
-	"k8s.io/client-go/util/keyutil"
 	componentbaseconfig "k8s.io/component-base/config"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
+// bootstrapKubeconfig retrieves an already existing kubeconfig for the Garden cluster from the Seed or bootstraps a new one
 func bootstrapKubeconfig(
 	ctx context.Context,
 	logger *logrus.Logger,
-	gardenClientConnection *config.GardenClientConnection,
+	gardenClientConnection config.GardenClientConnection,
 	seedClientConnection componentbaseconfig.ClientConnectionConfiguration,
 	seedConfig *config.SeedConfig,
 ) (
 	[]byte,
-	string,
-	string,
+	*string,
+	*string,
 	error,
 ) {
 	seedRESTCfg, err := kubernetes.RESTConfigFromClientConnectionConfiguration(&seedClientConnection, nil)
 	if err != nil {
-		return nil, "", "", err
+		return nil, nil, nil, err
 	}
-	k8sSeedClient, err := kubernetes.NewWithConfig(kubernetes.WithRESTConfig(seedRESTCfg))
+
+	seedClient, err := kubernetes.NewWithConfig(kubernetes.WithRESTConfig(seedRESTCfg))
 	if err != nil {
-		return nil, "", "", err
+		return nil, nil, nil, err
 	}
 
-	kubeconfigSecret := &corev1.Secret{}
-	if err := k8sSeedClient.DirectClient().Get(ctx, kutil.Key(gardenClientConnection.KubeconfigSecret.Namespace, gardenClientConnection.KubeconfigSecret.Name), kubeconfigSecret); client.IgnoreNotFound(err) != nil {
-		return nil, "", "", err
+	gardenKubeconfig, err := bootstraputil.GetKubeconfigFromSecret(ctx, seedClient.Client(), gardenClientConnection.KubeconfigSecret.Namespace, gardenClientConnection.KubeconfigSecret.Name)
+	if err != nil {
+		return nil, nil, nil, err
 	}
 
-	if len(kubeconfigSecret.Data[kubernetes.KubeConfig]) > 0 {
+	if gardenKubeconfig != nil {
```

Comment:
Can you move the length check from `bootstraputil.GetKubeconfigFromSecret` to this `if`-clause?
```suggestion
	if len(gardenKubeconfig) > 0 {
```